api/client/src/main/java/org/projectnessie/client/auth/oauth2/TokenExchangeConfig.java

```diff
@@ -71,19 +72,37 @@ static TokenExchangeConfig fromConfigSupplier(Function<String, String> config) {
     applyConfigOption(config, CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_SCOPES, builder::scope);
     applyConfigOption(config, CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_AUDIENCE, builder::audience);
     String subjectToken = config.apply(CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_SUBJECT_TOKEN);
+    String actorToken = config.apply(CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_ACTOR_TOKEN);
+    AtomicReference<URI> subjectTokenType = new AtomicReference<>(URN_ACCESS_TOKEN);
+    AtomicReference<URI> actorTokenType = new AtomicReference<>(URN_ACCESS_TOKEN);
+    applyConfigOption(
+        config,
+        CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_SUBJECT_TOKEN_TYPE,
+        subjectTokenType::set,
+        URI::create);
+    applyConfigOption(
+        config,
+        CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_ACTOR_TOKEN_TYPE,
+        actorTokenType::set,
+        URI::create);
+    // If a subject token is statically provided, use it. If no subject token is provided, then let
+    // the client be the subject, since a subject token is always required.
     if (subjectToken != null) {
-      String subjectTokenType = config.apply(CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_SUBJECT_TOKEN_TYPE);
-      builder.subjectToken(
-          TypedToken.of(
-              subjectToken,
-              subjectTokenType == null ? URN_ACCESS_TOKEN : URI.create(subjectTokenType)));
+      builder.subjectToken(TypedToken.of(subjectToken, subjectTokenType.get()));
+    } else {
+      builder.subjectTokenProvider(
+          (accessToken, refreshToken) ->
+              TypedToken.of(accessToken.getPayload(), subjectTokenType.get()));
     }
-    String actorToken = config.apply(CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_ACTOR_TOKEN);
+    // If an actor token is statically provided, use it. If no actor token is provided, but a
+    // subject token is statically provided, then let the client be the actor; otherwise, no actor
+    // token should be used.
     if (actorToken != null) {
-      String actorTokenType = config.apply(CONF_NESSIE_OAUTH2_TOKEN_EXCHANGE_ACTOR_TOKEN_TYPE);
-      builder.actorToken(
-          TypedToken.of(
-              actorToken, actorTokenType == null ? URN_ACCESS_TOKEN : URI.create(actorTokenType)));
+      builder.actorToken(TypedToken.of(actorToken, actorTokenType.get()));
+    } else if (subjectToken != null) {
```

Comment:
We check `subjectToken` but do not use it inside the `if`?

Reply:
Yes, the idea is that if a static subject token was provided, the client should use its own token as the actor token.

Reply:
noted :)